File: src/main/downloads/segmentedDownloader.test.ts
import { existsSync } from 'node:fs'
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises'
import { join } from 'node:path'
import { tmpdir } from 'node:os'
import { createHash } from 'node:crypto'
import { afterEach, describe, expect, it } from 'vitest'
import { SegmentedDownloader } from './segmentedDownloader'
import { createHttpDownloadServer } from '../../test/httpDownloadServer'

type TempPaths = {
  dir: string
  filePath: string
  tempPath: string
  metaPath: string
}

async function makeTempPaths(name: string): Promise<TempPaths> {
  const dir = await mkdtemp(join(tmpdir(), `arus-${name}-`))
  const filePath = join(dir, 'download.bin')
  return {
    dir,
    filePath,
    tempPath: `${filePath}.part`,
    metaPath: `${filePath}.part.meta.json`
  }
}

describe('SegmentedDownloader', () => {
  const cleanups: Array<() => Promise<void>> = []

  afterEach(async () => {
    while (cleanups.length > 0) {
      const cleanup = cleanups.pop()
      await cleanup?.()
    }
  })

  it('downloads with dynamic segmentation and produces byte-identical output', async () => {
    const body = Buffer.alloc(2 * 1024 * 1024, 0)
    for (let i = 0; i < body.length; i += 1) {
      body[i] = i % 251
    }

    const server = await createHttpDownloadServer({
      body,
      supportRanges: true,
      etag: '"v1-dynamic"',
      contentDisposition: 'attachment; filename="payload.bin"'
    })
    const paths = await makeTempPaths('segmented')
    cleanups.push(async () => {
      await server.close()
      await rm(paths.dir, { recursive: true, force: true })
    })

    const seenSegments: number[] = []
    const downloader = new SegmentedDownloader({
      id: 'task-1',
      url: server.url,
      filePath: paths.filePath,
      tempPath: paths.tempPath,
      metaPath: paths.metaPath,
      connections: 4,
      minSegmentSizeBytes: 128 * 1024,
      maxRetries: 2,
      onProgress: (progress) => {
        if (progress.segments) {
          seenSegments.push(progress.segments.length)
        }
      }
    })

    await downloader.start()

    const result = await readFile(paths.filePath)
    expect(result.equals(body)).toBe(true)
    expect(Math.max(...seenSegments, 1)).toBeGreaterThan(1)
    expect(existsSync(paths.tempPath)).toBe(false)
    expect(existsSync(paths.metaPath)).toBe(false)
    expect(existsSync(`${paths.metaPath}.tmp`)).toBe(false)

    const rangeGets = server.requests.filter(
      (request) => request.method === 'GET' && request.range
    )
    expect(rangeGets.length).toBeGreaterThan(1)
  })

  it('recreates a destination folder that disappeared before the task resumed', async () => {
    const body = Buffer.from('folder-recovery-payload')
    const server = await createHttpDownloadServer({ body, supportRanges: false })
    const root = await mkdtemp(join(tmpdir(), 'arus-missing-folder-'))
    const directory = join(root, 'deleted', 'Downloads')
    const filePath = join(directory, 'download.bin')
    const paths: TempPaths = {
      dir: root,
      filePath,
      tempPath: `${filePath}.part`,
      metaPath: `${filePath}.part.meta.json`
    }
    cleanups.push(async () => {
      await server.close()
      await rm(paths.dir, { recursive: true, force: true })
    })

    const downloader = new SegmentedDownloader({
      id: 'task-missing-folder',
      url: server.url,
      filePath: paths.filePath,
      tempPath: paths.tempPath,
      metaPath: paths.metaPath,
      connections: 8,
      onProgress: () => undefined
    })

    await downloader.start()

    expect(await readFile(paths.filePath)).toEqual(body)
    expect(existsSync(directory)).toBe(true)
  })

  it('retries a dropped single-stream connection and resumes the partial file', async () => {
    const body = Buffer.alloc(128 * 1024, 4)
    for (let i = 0; i < body.length; i += 1) {
      body[i] = (i * 11) % 256
    }

    const server = await createHttpDownloadServer({
      body,
      supportRanges: false,
      failFullGets: 1
    })
    const paths = await makeTempPaths('single-retry')
    cleanups.push(async () => {
      await server.close()
      await rm(paths.dir, { recursive: true, force: true })
    })

    const downloader = new SegmentedDownloader({
      id: 'task-single-retry',
      url: server.url,
      filePath: paths.filePath,
      tempPath: paths.tempPath,
      metaPath: paths.metaPath,
      connections: 8,
      maxRetries: 1,
      onProgress: () => undefined
    })

    await downloader.start()

    expect(await readFile(paths.filePath)).toEqual(body)
    expect(server.requests.filter((request) => request.method === 'GET' && !request.range)).toHaveLength(2)
  })

  it('resumes from persisted meta without re-downloading completed bytes', async () => {
    const body = Buffer.alloc(1024 * 1024, 7)
    for (let i = 0; i < body.length; i += 1) {
      body[i] = (i * 3) % 256
    }

    const server = await createHttpDownloadServer({
      body,
      supportRanges: true,
      etag: '"resume-etag"',
      lastModified: 'Wed, 01 Jan 2025 00:00:00 GMT'
    })
    const paths = await makeTempPaths('resume')
    cleanups.push(async () => {
      await server.close()
      await rm(paths.dir, { recursive: true, force: true })
    })

    // Seed a partial preallocated file + versioned checkpoint as if we paused mid-download.
    const half = Math.floor(body.length / 2)
    const part = Buffer.alloc(body.length, 0)
    body.copy(part, 0, 0, half)
    await writeFile(paths.tempPath, part)
    await writeFile(
      paths.metaPath,
      JSON.stringify({
        version: 2,
        url: server.url,
        totalBytes: body.length,
        connections: 4,
        minSegmentSizeBytes: 64 * 1024,
        segments: [
          { index: 0, start: 0, end: half - 1, downloaded: half },
          { index: 1, start: half, end: body.length - 1, downloaded: 0 }
        ],
        etag: '"resume-etag"',
        lastModified: 'Wed, 01 Jan 2025 00:00:00 GMT',
        contentMd5: null,
        contentDisposition: null,
        mode: 'segmented'
      })
    )

    const requestsBeforeResume = server.requests.length

    const downloader = new SegmentedDownloader({
      id: 'task-resume',
      url: server.url,
      filePath: paths.filePath,
      tempPath: paths.tempPath,
      metaPath: paths.metaPath,
      connections: 4,
      minSegmentSizeBytes: 64 * 1024,
      maxRetries: 2,
      onProgress: () => undefined
    })
    await downloader.start()

    const result = await readFile(paths.filePath)
    expect(result.equals(body)).toBe(true)
    expect(existsSync(paths.tempPath)).toBe(false)
    expect(existsSync(paths.metaPath)).toBe(false)

    const resumeRanges = server.requests
      .slice(requestsBeforeResume)
      .filter((request) => request.method === 'GET' && request.range)
    expect(resumeRanges.length).toBeGreaterThan(0)
    // Resumed work should request bytes from the unfinished half (not restart at 0).
    expect(
      resumeRanges.some((request) => {
        const match = /^bytes=(\d+)-/.exec(request.range || '')
        return match ? Number(match[1]) >= half : false
      })
    ).toBe(true)
    // Completed first half must not be requested again from byte 0 as a full-file range.
    expect(
      resumeRanges.every((request) => {
        const match = /^bytes=(\d+)-/.exec(request.range || '')
        if (!match) {
          return true
        }
        const start = Number(match[1])
        // Probe uses 0-0; segmented resume for incomplete work starts at >= half.
        return start === 0 || start >= half
      })
    ).toBe(true)
  })

  it('falls back to a single connection when the server rejects ranges', async () => {
    const body = Buffer.from('hello-arus-single-stream-payload-0123456789')
    const server = await createHttpDownloadServer({
      body,
      supportRanges: false,
      rejectHead: true
    })
    const paths = await makeTempPaths('norange')
    cleanups.push(async () => {
      await server.close()
      await rm(paths.dir, { recursive: true, force: true })
    })

    let maxSegments = 0
    const downloader = new SegmentedDownloader({
      id: 'task-single',
      url: server.url,
      filePath: paths.filePath,
      tempPath: paths.tempPath,
      metaPath: paths.metaPath,
      connections: 8,
      minSegmentSizeBytes: 64 * 1024,
      onProgress: (progress) => {
        maxSegments = Math.max(maxSegments, progress.segments?.length ?? 0)
      }
    })

    await downloader.start()

    const result = await readFile(paths.filePath)
    expect(result.equals(body)).toBe(true)
    expect(maxSegments).toBe(0)

    const rangeGets = server.requests.filter(
      (request) => request.method === 'GET' && request.range && request.range !== 'bytes=0-0'
    )
    // Probe may use bytes=0-0; after fallback no multi-range segmented GETs.
    expect(rangeGets.length).toBe(0)
  })

  it('uses single-stream when HEAD advertises ranges but Range GETs return 403', async () => {
    const body = Buffer.alloc(256 * 1024, 9)
    for (let i = 0; i < body.length; i += 1) {
      body[i] = (i * 7) % 256
    }

    const server = await createHttpDownloadServer({
      body,
      supportRanges: true,
      rejectRangeWithStatus: 403,
      etag: '"anti-leech"'
    })
    const paths = await makeTempPaths('antihotlink')
    cleanups.push(async () => {
      await server.close()
      await rm(paths.dir, { recursive: true, force: true })
    })

    let maxSegments = 0
    const downloader = new SegmentedDownloader({
      id: 'task-403-range',
      url: server.url,
      filePath: paths.filePath,
      tempPath: paths.tempPath,
      metaPath: paths.metaPath,
      connections: 8,
      minSegmentSizeBytes: 32 * 1024,
      onProgress: (progress) => {
        maxSegments = Math.max(maxSegments, progress.segments?.length ?? 0)
      }
    })

    await downloader.start()

    const result = await readFile(paths.filePath)
    expect(result.equals(body)).toBe(true)
    expect(maxSegments).toBe(0)

    // Full-body GET (no Range) must succeed after probe Range was rejected.
    const fullGets = server.requests.filter(
      (request) => request.method === 'GET' && !request.range
    )
    expect(fullGets.length).toBeGreaterThan(0)
    expect(fullGets.some((request) => request.headers['user-agent'])).toBe(true)
    expect(fullGets.some((request) => request.headers.referer)).toBe(true)
  })

  it('validates Content-MD5 when the server provides it', async () => {
    const body = Buffer.from('checksum-me-please')
    const md5 = createHash('md5').update(body).digest('base64')
    const server = await createHttpDownloadServer({
      body,
      supportRanges: true,
      contentMd5: md5,
      etag: '"md5-file"'
    })
    const paths = await makeTempPaths('md5')
    cleanups.push(async () => {
      await server.close()
      await rm(paths.dir, { recursive: true, force: true })
    })

    const downloader = new SegmentedDownloader({
      id: 'task-md5',
      url: server.url,
      filePath: paths.filePath,
      tempPath: paths.tempPath,
      metaPath: paths.metaPath,
      connections: 4,
      minSegmentSizeBytes: 8,
      onProgress: () => undefined
    })

    await downloader.start()
    expect(await readFile(paths.filePath)).toEqual(body)
  })
})

File: test/segmented_downloader_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import 'package:arus_flutter/src/core/path_utils.dart';
import 'package:arus_flutter/src/services/segmented_downloader.dart';

void main() {
  test('downloads a byte-identical segmented file into a missing directory', () async {
    final bytes = List<int>.generate(96 * 1024, (index) => (index * 31) % 251);
    final server = await _serve(bytes, supportsRanges: true);
    final root = await Directory.systemTemp.createTemp('arus-segmented-test-');
    final target = joinPath(joinPath(root.path, 'nested', 'folder'), 'sample.bin');

    try {
      final downloader = SegmentedDownloader(
        id: 'test-segmented',
        url: 'http://127.0.0.1:${server.port}/sample.bin',
        filePath: target,
        tempPath: '$target.part',
        metaPath: '$target.part.meta.json',
        connections: 4,
        minSegmentSizeBytes: 8 * 1024,
        maxRetries: 2,
        onProgress: (_) {},
      );

      await downloader.start();

      expect(await File(target).readAsBytes(), bytes);
      expect(await File(target).exists(), isTrue);
      expect(await File('$target.part').exists(), isFalse);
      expect(await File('$target.part.meta.json').exists(), isFalse);
    } finally {
      await server.close(force: true);
      await root.delete(recursive: true);
    }
  });

  test('resumes a previous segmented checkpoint and removes it after assembly', () async {
    final bytes = List<int>.generate(96 * 1024, (index) => (index * 13) % 239);
    final server = await _serve(bytes, supportsRanges: true);
    final root = await Directory.systemTemp.createTemp('arus-resume-test-');
    final target = joinPath(root.path, 'resume.bin');
    final half = bytes.length ~/ 2;
    final part = List<int>.filled(bytes.length, 0);
    part.setRange(0, half, bytes, 0);

    try {
      await File('$target.part').create(recursive: true);
      await File('$target.part').writeAsBytes(part);
      await File('$target.part.meta.json').writeAsString(
        jsonEncode(<String, dynamic>{
          'version': 2,
          'mode': 'segmented',
          'url': 'http://127.0.0.1:${server.port}/resume.bin',
          'totalBytes': bytes.length,
          'segments': <Map<String, int>>[
            <String, int>{'index': 0, 'start': 0, 'end': half - 1, 'downloaded': half},
            <String, int>{'index': 1, 'start': half, 'end': bytes.length - 1, 'downloaded': 0},
          ],
        }),
      );

      final downloader = SegmentedDownloader(
        id: 'test-resume',
        url: 'http://127.0.0.1:${server.port}/resume.bin',
        filePath: target,
        tempPath: '$target.part',
        metaPath: '$target.part.meta.json',
        connections: 4,
        minSegmentSizeBytes: half,
        maxRetries: 2,
        onProgress: (_) {},
      );

      await downloader.start();

      expect(await File(target).readAsBytes(), bytes);
      expect(await File('$target.part.meta.json').exists(), isFalse);
    } finally {
      await server.close(force: true);
      await root.delete(recursive: true);
    }
  });

  test('falls back to a single stream when the server ignores ranges', () async {
    final bytes = List<int>.generate(20 * 1024, (index) => index % 199);
    final server = await _serve(bytes, supportsRanges: false);
    final root = await Directory.systemTemp.createTemp('arus-single-test-');
    final target = joinPath(root.path, 'single.bin');

    try {
      final downloader = SegmentedDownloader(
        id: 'test-single',
        url: 'http://127.0.0.1:${server.port}/single.bin',
        filePath: target,
        tempPath: '$target.part',
        metaPath: '$target.part.meta.json',
        connections: 8,
        minSegmentSizeBytes: 4 * 1024,
        maxRetries: 1,
        onProgress: (_) {},
      );

      await downloader.start();

      expect(await File(target).readAsBytes(), bytes);
      expect(await File('$target.part').exists(), isFalse);
      expect(await File('$target.part.meta.json').exists(), isFalse);
    } finally {
      await server.close(force: true);
      await root.delete(recursive: true);
    }
  });

  test('rebuilds successfully when Range support disappears after the probe', () async {
    final bytes = List<int>.generate(128 * 1024, (index) => (index * 23) % 241);
    final server = await _serveRangeLoss(bytes);
    final root = await Directory.systemTemp.createTemp('arus-range-loss-test-');
    final target = joinPath(root.path, 'range-loss.bin');

    try {
      final downloader = SegmentedDownloader(
        id: 'test-range-loss',
        url: 'http://127.0.0.1:${server.port}/range-loss.bin',
        filePath: target,
        tempPath: '$target.part',
        metaPath: '$target.part.meta.json',
        connections: 4,
        minSegmentSizeBytes: 8 * 1024,
        maxRetries: 2,
        onProgress: (_) {},
      );

      await downloader.start();

      expect(await File(target).readAsBytes(), bytes);
      expect(await File('$target.part').exists(), isFalse);
      expect(await File('$target.part.meta.json').exists(), isFalse);
    } finally {
      await server.close(force: true);
      await root.delete(recursive: true);
    }
  });

  test('retries a dropped single stream and resumes the partial file', () async {
    final bytes = List<int>.generate(48 * 1024, (index) => (index * 17) % 233);
    final server = await _serveWithDrop(bytes);
    final root = await Directory.systemTemp.createTemp('arus-retry-test-');
    final target = joinPath(root.path, 'retry.bin');

    try {
      final downloader = SegmentedDownloader(
        id: 'test-retry',
        url: 'http://127.0.0.1:${server.port}/retry.bin',
        filePath: target,
        tempPath: '$target.part',
        metaPath: '$target.part.meta.json',
        connections: 8,
        minSegmentSizeBytes: 4 * 1024,
        maxRetries: 2,
        onProgress: (_) {},
      );

      await downloader.start();

      expect(await File(target).readAsBytes(), bytes);
      expect(await File('$target.part').exists(), isFalse);
      expect(await File('$target.part.meta.json').exists(), isFalse);
    } finally {
      await server.close(force: true);
      await root.delete(recursive: true);
    }
  });
}

Future<HttpServer> _serve(List<int> bytes, {required bool supportsRanges}) async {
  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
  server.listen((request) async {
    final response = request.response;
    response.headers.set(HttpHeaders.etagHeader, '"arus-test"');

    if (request.method == 'HEAD') {
      response.statusCode = HttpStatus.ok;
      response.contentLength = bytes.length;
      await response.close();
      return;
    }

    final range = request.headers.value(HttpHeaders.rangeHeader);
    if (supportsRanges && range != null) {
      final match = RegExp(r'bytes=(\d+)-(\d+)?').firstMatch(range);
      if (match != null) {
        final start = int.parse(match.group(1)!);
        final requestedEnd = match.group(2) == null ? bytes.length - 1 : int.parse(match.group(2)!);
        final end = requestedEnd.clamp(start, bytes.length - 1).toInt();
        response.statusCode = HttpStatus.partialContent;
        response.headers.set(HttpHeaders.contentRangeHeader, 'bytes $start-$end/${bytes.length}');
        response.contentLength = end - start + 1;
        response.add(bytes.sublist(start, end + 1));
        await response.close();
        return;
      }
    }

    response.statusCode = HttpStatus.ok;
    response.contentLength = bytes.length;
    response.add(bytes);
    await response.close();
  });
  return server;
}

Future<HttpServer> _serveWithDrop(List<int> bytes) async {
  var dropped = false;
  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
  server.listen((request) async {
    final response = request.response;
    response.headers.set(HttpHeaders.etagHeader, '"arus-retry"');

    if (request.method == 'HEAD') {
      response.statusCode = HttpStatus.ok;
      response.contentLength = bytes.length;
      await response.close();
      return;
    }

    final range = request.headers.value(HttpHeaders.rangeHeader);
    final match = range == null ? null : RegExp(r'bytes=(\d+)-').firstMatch(range);
    final start = match == null ? 0 : int.parse(match.group(1)!);

    if (range != null && start > 0) {
      response.statusCode = HttpStatus.partialContent;
      response.headers.set(HttpHeaders.contentRangeHeader, 'bytes $start-${bytes.length - 1}/${bytes.length}');
      response.contentLength = bytes.length - start;
      response.add(bytes.sublist(start));
      await response.close();
      return;
    }

    response.statusCode = HttpStatus.ok;
    if (!dropped) {
      dropped = true;
      response.contentLength = -1;
      response.add(bytes.sublist(0, bytes.length ~/ 2));
    } else {
      response.contentLength = bytes.length;
      response.add(bytes);
    }
    await response.close();
  });
  return server;
}

Future<HttpServer> _serveRangeLoss(List<int> bytes) async {
  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
  var probeSeen = false;
  server.listen((request) async {
    final response = request.response;
    response.headers.set(HttpHeaders.etagHeader, '"arus-range-loss"');

    if (request.method == 'HEAD') {
      response.statusCode = HttpStatus.ok;
      response.contentLength = bytes.length;
      await response.close();
      return;
    }

    final range = request.headers.value(HttpHeaders.rangeHeader);
    if (range == 'bytes=0-0' && !probeSeen) {
      probeSeen = true;
      response.statusCode = HttpStatus.partialContent;
      response.headers.set(HttpHeaders.contentRangeHeader, 'bytes 0-0/${bytes.length}');
      response.contentLength = 1;
      response.add(<int>[bytes.first]);
      await response.close();
      return;
    }

    if (range != null) {
      response.statusCode = HttpStatus.forbidden;
      response.contentLength = 0;
      await response.close();
      return;
    }

    response.statusCode = HttpStatus.ok;
    response.contentLength = bytes.length;
    response.add(bytes);
    await response.close();
  });
  return server;
}
